File: backends/php-cakephp/src/Controller/ProxyController.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\Controller;

use Psr\Http\Message\ResponseInterface;

/**
 * 認可サーバーへのリバースプロキシを提供するコントローラークラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 * @package App\Controller
 */
class ProxyController extends AppController
{
    private string $authServerUrl;

    /**
     * 初期化処理を行います。
     *
     * @return void
     */
    public function initialize(): void
    {
        parent::initialize();
        $this->autoRender = false;
        $this->authServerUrl = env('AUTH_SERVER_URL', 'http://host.docker.internal:8080/function/php');
    }

    /**
     * 指定パスへ GET リクエストを転送し、認可サーバーのレスポンスをそのまま返します。
     *
     * @param string $path 転送先パス
     * @return ResponseInterface レスポンス
     */
    private function proxyGet(string $path): ResponseInterface
    {
        $query = $this->request->getUri()->getQuery();
        $url = $this->authServerUrl . '/' . $path . ($query ? '?' . $query : '');

        $headers = ['Accept: application/json'];
        $auth = $this->request->getHeaderLine('Authorization');
        if ($auth) {
            $headers[] = 'Authorization: ' . $auth;
        }

        $context = stream_context_create([
            'http' => [
                'method'        => 'GET',
                'header'        => implode("\r\n", $headers),
                'ignore_errors' => true,
            ],
        ]);

        $body = @file_get_contents($url, false, $context);
        if ($body === false) {
            return $this->response->withStatus(502)->withType('application/json')
                ->withStringBody('{"error":"Failed to connect to auth server"}');
        }

        $statusCode = 200;
        if (isset($http_response_header[0])) {
            preg_match('/HTTP\/[\d.]+ (\d+)/', $http_response_header[0], $m);
            $statusCode = (int)($m[1] ?? 200);
        }

        return $this->response->withStatus($statusCode)->withType('application/json')
            ->withStringBody($body);
    }

    /**
     * ヘルスチェック応答を返します。
     *
     * @return ResponseInterface レスポンス
     */
    public function health(): ResponseInterface
    {
        return $this->response->withStatus(200)->withType('application/json')
            ->withStringBody('{"status":"ok"}');
    }

    /**
     * クライアント一覧を認可サーバーから取得して返します。
     *
     * @return ResponseInterface レスポンス
     */
    public function clients(): ResponseInterface
    {
        return $this->proxyGet('api/clients');
    }

    /**
     * クライアント会員向け JWT を発行して返します。
     *
     * @return ResponseInterface レスポンス
     */
    public function gateIssue(): ResponseInterface
    {
        return $this->proxyGet('api/gate/issue');
    }

    /**
     * JWT を検証してペイロードを返します。
     *
     * @return ResponseInterface レスポンス
     */
    public function gateVerify(): ResponseInterface
    {
        $identifier = $this->request->getParam('identifier', '');
        return $this->proxyGet('api/gate/client/' . $identifier . '/verify');
    }
}

File: backends/ruby-rails/app/controllers/proxy_controller.rb
# frozen_string_literal: true
#
# This is a program developed by BobTabo.
#
# Copyright (c) 2026 BobTabo. All Rights Reserved.

require 'net/http'
require 'uri'

# This is a program developed by BobTabo.
#
# Copyright (c) 2026 BobTabo. All Rights Reserved.
#
# 認可サーバーへのリバースプロキシを提供するコントローラークラスです。
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
class ProxyController < ApplicationController
  skip_before_action :verify_authenticity_token, raise: false

  # ヘルスチェック応答を返します。
  def health
    render json: { status: 'ok' }
  end

  # クライアント一覧を認可サーバーから取得して返します。
  def clients
    proxy_get('api/clients')
  end

  # クライアント会員向け JWT を発行して返します。
  def gate_issue
    proxy_get('api/gate/issue')
  end

  # JWT を検証してペイロードを返します。
  def gate_verify
    proxy_get("api/gate/client/#{params[:identifier]}/verify")
  end

  private

  # 認可サーバーの URL を取得します。
  def auth_server_url
    ENV['AUTH_SERVER_URL'] || 'http://host.docker.internal:8080/function/php'
  end

  # 指定パスへ GET リクエストを転送し、認可サーバーのレスポンスをそのまま返します。
  def proxy_get(path)
    query = request.query_string
    uri = URI("#{auth_server_url}/#{path}#{query.present? ? "?#{query}" : ''}")

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'
    req = Net::HTTP::Get.new(uri)
    req['Accept'] = 'application/json'
    auth = request.headers['Authorization']
    req['Authorization'] = auth if auth.present?

    resp = http.request(req)
    render plain: resp.body, status: resp.code.to_i, content_type: 'application/json'
  rescue StandardError => e
    render json: { error: e.message }, status: :bad_gateway
  end
end
